### PR TITLE
First version of minnie-kenny.sh BA-5810

### PR DESCRIPTION
**Pull request by kshakir**

### CODE REVIEW

File: codecov.yml [hunk 1]
```diff
@@ -0,0 +1,21 @@
+comment:
+  require_changes: yes
+
+# Keeping it 100! If you're looking to change that, either:
+# - Go add a test!
+# - Only for extreme cases: exclude the line from kcov https://www.mankier.com/1/kcov#--exclude-pattern
+coverage:
+  range: 100..100
```
Comment (mcovarr) [inline codecov.yml:8]: 💯

File: docs/Acknowledgments.md [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+## Online hosting / services
+- [Codecov](https://codecov.io/): Group, merge, archive, and compare coverage reports
+- [GitHub](https://github.com/features): All your code in one place
+- [Read the Docs](https://readthedocs.org): Simplifies software documentation
+- [Travis CI](https://docs.travis-ci.com/user/for-beginners/): Automatically building and testing code changes
+
+## JetBrains development environment
+- [IntelliJ IDEA](https://www.jetbrains.com/idea/features/): Making development an enjoyable experience
+- [BashSupport](https://plugins.jetbrains.com/plugin/4230-bashsupport): Bash language support for the IntelliJ platform
+- [Shellcheck](https://plugins.jetbrains.com/plugin/10195-shellcheck/): Analyze scripts with shellcheck
```
Comment (mcovarr) [inline docs/Acknowledgments.md:57]: Wow, fair warning I will likely be citing this acknowledgments doc. 🙂

File: docs/Install.md [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+```gitconfig
+# NOTE: The stanza [secrets] must be included to be a valid git-config file
+[secrets]
+    providers = git secrets --aws-provider
+    patterns = (A3T[A-Z0-9]|AKIA|AGPA|AIDA|AROA|AIPA|ANPA|ANVA|ASIA)[A-Z0-9]{16}
+    patterns = (\"|')?(AWS|aws|Aws)?_?(SECRET|secret|Secret)?_?(ACCESS|access|Access)?_?(KEY|key|Key)(\"|')?\\s*(:|=>|=)\\s*(\"|')?[A-Za-z0-9/\\+=]{40}(\"|')?
+    patterns = (\"|')?(AWS|aws|Aws)?_?(ACCOUNT|account|Account)_?(ID|id|Id)?(\"|')?\\s*(:|=>|=)\\s*(\"|')?[0-9]{4}\\-?[0-9]{4}\\-?[0-9]{4}(\"|')?
+    allowed = AKIAIOSFODNN7EXAMPLE
+    allowed = wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY
+    # NOTE: The above example is the equivalent of `git secrets --register-aws`. Futher customize for your own git repo.
```
Comment (mcovarr) [inline docs/Install.md]: Futher => Further

File: docs/js/extra.js [hunk 4]
```diff
@@ -0,0 +1,219 @@
+var nodemcu = nodemcu || {};
+(function () {
+  'use strict';
+  //var languageCodeToNameMap = {en: 'English', de: 'Deutsch'};
```
Comment (mcovarr) [inline docs/js/extra.js:4]: ?

File: minnie-kenny.sh [hunk 5]
```diff
@@ -0,0 +1,189 @@
+#!/bin/sh
+#   Use this script to ensure git-secrets are setup
+#   https://minnie-kenny.readthedocs.io/
+
+set -eu # -o pipefail isn't supported by POSIX
```
Comment (mcovarr) [inline minnie-kenny.sh:5]: 🙁

File: minnie-kenny.sh [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+
+minnie_kenny_command_name=${0##*/}
```
Comment (mcovarr) [inline minnie-kenny.sh:7]: 😎

File: minnie-kenny.sh [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+}
+
+echo_out() { if [ ${minnie_kenny_quiet} -ne 1 ]; then echo "$@"; fi; }
+echo_err() { if [ ${minnie_kenny_quiet} -ne 1 ]; then echo "$@" 1>&2; fi; }
+
+process_arguments() {
+  while [ $# -gt 0 ]; do
+    case "$1" in
+      -q | --quiet)
+        minnie_kenny_quiet=1
```
Comment (mcovarr) [inline minnie-kenny.sh:36]: Is there a reason not to use `true` and `false` if these represent booleans?

File: docs/CommonQuestions.md [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+baz
+$
+```
+
+`git secrets` already requires `bash` to run, so adding GNU's grep shouldn't be much more of an addition.
+
+## How do I uninstall `minnie-kenny.sh` / `git secrets`?
+
+`minnie-kenny.sh` invokes `git secrets --install` to add the github hooks, and configures the `.git/config` to read from
+`minnie-kenny.gitconfg`. To remove the configurations that `minnnie-kenny.sh` has installed:
```
Comment (mcovarr) [inline docs/CommonQuestions.md]: minnie-kenny.gitconfg => minnie-kenny.gitconfig

File: docs/CommonQuestions.md [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+
+- edit the `.git/config`:
+    - remove the `[secrets]` stanza
+    - remove the `[include]` for `minnie-kenny.sh`
+- remove any calls to `git secrets` in the hooks:
+    - `.git/hooks/commit-msg`
+    - `.git/hooks/pre-commit`
+    - `.git/hooks/prepare-commit-msg`
+
+To remove the `git-secrets` from your system:
```
Comment (mcovarr) [inline docs/CommonQuestions.md]: remove these things?

**Review by danbills — DISMISSED** [review]
wow, looks great!

Small Disclaimer:
I don't have enough bash experience to have strong opinions about style or idioms.  

My one suggestion is that you run it through shell check to see if it has recommendations.

File: docs/Contributing.md [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+[MkDocs](https://www.mkdocs.org), and may rendered locally using `mkdocs serve`.
+
+## Pull requests
+
+Updates are welcome and may be submitted via [GitHub pull
+requests](https://github.com/broadinstitute/minnie-kenny/pulls).
+
+Every pull request should refer to a Jira issue in the description. See above under [Issue Tracking](#issue-tracking)
+for information on Jira. The pull request description should contain the string "BA-#" where "#" is the Jira issue
+number, for example "My latest pull request BA-99999".
```
Comment (mcovarr) [inline docs/Contributing.md:42]: does it need to be surrounded by square brackets to work or am I (mis)remembering from a previous JIRA "integration"?

File: docs/Usage.md [hunk 11]
```diff
@@ -0,0 +1,39 @@
+## Overview
+
+```console
+Usage:
+    minnie-kenny.sh
+    -f | --force                Modify the git config to run git secrets
+    -n | --no-force             Do not modify the git config only verify installation
```
Comment (mcovarr) [inline docs/Usage.md]: config, only

File: test/minnie-kenny.bats [hunk 12]
```diff
@@ -0,0 +1,282 @@
...
+    -n | --no-force             Do not modify the git config only verify installation
+    -s | --strict               Require git-secrets to be setup or fail
+    -q | --quiet                Do not output any status messages
+    -i | --include=FILE         Path to the include for git-config (default: \"minnie-kenny.gitconfig\")"
+
+  # The minnie kenny usage message with a leading newline
+  minnie_kenny_message_usage=$'\n'"${minnie_kenny_message_usage_trimmed}"
+
+  # The formatted check mark at the beginning of the install message
+  minnie_kenny_check_mark="$(tput setaf 2)✓$(tput sgr 0)"
```
Comment (mcovarr) [inline test/minnie-kenny.bats:48]: what the

**Review by mcovarr — DISMISSED** [review]
Awesome!

File: docs/js/extra.js (hunk 4 above)
Comment (kshakir) [reply docs/js/extra.js:4]: Copypasta bit-for-bit from Cromwell: https://raw.githubusercontent.com/broadinstitute/cromwell/develop/docs/js/extra.js

Not really sure where to document this, as putting the attribution inside the document would mean I couldn't just copy it again in the future.

I did track down what may be the "original" though. The contents of this extra.js seem to be borrowed from the history of https://github.com/nodemcu/nodemcu-firmware/commits/master/docs/js/extra.js. I referred to that file's git history in the current Acknowledgments docs.

File: minnie-kenny.sh (hunk 6 above)
Comment (kshakir) [reply minnie-kenny.sh:7]: Borrowed from https://github.com/vishnubob/wait-for-it/blob/54d1f0bfeb6557adf8a3204455389d0901652242/wait-for-it.sh#L4

File: minnie-kenny.sh (hunk 7 above)
Comment (kshakir) [reply minnie-kenny.sh:36]: Both [git-secrets](https://github.com/awslabs/git-secrets/blob/3958dacceeebeab84e2a3c686c00fb9bde17cb55/git-secrets#L331-L335) and [wait-for-it.sh](https://github.com/vishnubob/wait-for-it/blob/54d1f0bfeb6557adf8a3204455389d0901652242/wait-for-it.sh#L55) use this convention of `1` is true and `0` is false. Those two consistent practical-implementations overrode the various [theoretical-discussions on Stack Overflow](https://stackoverflow.com/questions/2953646/how-to-declare-and-use-boolean-variables-in-shell-script).

In the S.O. thread above, it is interestingly pointed out that the command `true` returns an exit code of `0`, while `false` returns a `1`. However those aren't variables, they're POSIX commands and the `0` and `1` are their respective exit statuses.

It seems that there are no POSIX/bash variables for `true` or `false`. Alternatively I could have used the strings `"true"` and `"false"`, but those seemed verbose and didn't match what git-secrets and wait-for-it.sh did for booleans. Note: I did examine the string-booleans for the `git rev-parse`, as those are the strings [output by the `git rev-parse` command](https://git-scm.com/docs/git-rev-parse#Documentation/git-rev-parse.txt---is-inside-work-tree)... assuming the `git` executable is found. 😦

**TL;DR Instead of `true`/`false` I could use `"true"`/`"false"` but those strings don't seem to be canonical.**

File: docs/CommonQuestions.md (hunk 8 above)
Comment (kshakir) [reply docs/CommonQuestions.md]: Ooops

File: docs/Contributing.md (hunk 10 above)
Comment (kshakir) [reply docs/Contributing.md:42]: As far as I can tell, the square brackets around Jira issue links started because one (or more?) people misread a confusing section of the Jira docs circa 2010, and then everyone else followed along...

[These old docs state](https://confluence.atlassian.com/display/APPLINKS030/Using+Link+Rendering#space-menu-link-content):

> You could then reference an issue in your JIRA project ... by just typing the issue key, e.g. [JRA-1234]

That sentence uses square brackets for some reason... but just below that line there's a table that explicitly mentions:

| To link to ...                     | Type this                       | Examples                  |
|------------------------------------|---------------------------------|---------------------------|
| an issue                           | \<Issue Key\>                   | _TEST-123_<br/>_ACME-456_ |
| a wiki page in the same project    | \[\<Page Name\>\]               | _\[My Draft Page\]_       |
| a wiki page in a different project | \[\<Space Key\>:\<Page Name\>\] | _\[TEST:My Draft Page\]_  |

In the current docs the examples are even clearer, [in this case for explicit commit messages](https://confluence.atlassian.com/display/APPLINKS060/Streamline+software+development+with+Atlassian+tools?preview=/831660181/831660193/Stream-commit-details3.png), where none of the example descriptions use square brackets yet still are detected.

**TL;DR I don't use brackets, as in this current PR, and linking still works.**

**Comment by kshakir** [comment]
I think I fixed everything mentioned:

- mentioned in this PR
- so far in both https://github.com/DataBiosphere/job-manager/pull/681
- so far in https://github.com/broadinstitute/cromwell/pull/5156
- bonus: fixed an issue where `-q` wasn't as-quiet-as-possible.

**Review by mcovarr — APPROVED** [review]

**Review by danbills — APPROVED** [review]
